File: neet_code_150_cpp/stack/min_stack.cpp
#include <vector>
#include <stack>
#include <iostream>

// AC
// しかしメモリ効率が悪い
class MinStack {
#define DEBUG
    private:
        std::vector<int> volume;
        std::vector<int> min_list;

    public:
        MinStack() {}

        void push(int val){
            if(min_list.empty()) {
                min_list.push_back(val);
            } else if(val <= min_list.back()) {
                min_list.push_back(val);
            }            

            volume.push_back(val);
        }

        void pop() {
            if(volume.back() == min_list.back()) {
                min_list.pop_back();
            }
            volume.pop_back();
        }

        int top() {
#ifdef DEBUG
            std::cout << volume.back() << std::endl;
#endif
            return volume.back();
        }

        int get_min() {
#ifdef DEBUG
            std::cout << min_list.back() << std::endl;
#endif
            return min_list.back();
        }

#ifdef DEBUG
        void show(void) {
            std::cout << "volume" << std::endl;
            for(const auto& v: volume) {
                std::cout << v << " ";
            }
            std::cout << std::endl;

            std::cout << "min_list" << std::endl;
            for(const auto& v: min_list) {
                std::cout << v << " ";
            }
            std::cout << std::endl;
        }
#endif
};

// AC
// メモリはあまり変わらない?
class MinStackB {
    private:
        std::vector<std::pair<int, int>> volume;
    public:
        MinStackB() {}

        void push(int val) {
            if(volume.empty()) {
                volume.push_back({val, val});
                return;
            }

            int current_min = volume.back().second;
            if(val <= current_min) {
                volume.push_back({val, val});
            } else {
                volume.push_back({val, current_min});
            }
        }

        void pop() {
            volume.pop_back();
        }

        int top() {
#ifdef DEBUG
            std::cout << volume.back().first << std::endl;
#endif
            return volume.back().first;
        }

        int get_min() {
#ifdef DEBUG
            std::cout << volume.back().second << std::endl;
#endif
            return volume.back().second;
        }
};

// 模範解答
class MinStackAns {
    public:
        MinStackAns() {
        }

        void push(int val) {
            stk.push(val);

            if (minStk.empty() || val < minStk.top().first) {
                minStk.push({val, 1});
            } else if (val == minStk.top().first) {
                minStk.top().second++;
            }
        }

        void pop() {
            if (stk.top() == minStk.top().first) {
                minStk.top().second--;
                if (minStk.top().second == 0) {
                    minStk.pop();
                }
            }
            stk.pop();
        }

        int top() {
            return stk.top();
        }

        int getMin() {
            return minStk.top().first;
        }
    private:
        std::stack<int> stk;
        std::stack<std::pair<int, int>> minStk;
};


int main(void) {
    MinStack ms_1;
    std::cout << "ms_1" << std::endl;

    ms_1.push(0);
    ms_1.push(1);
    ms_1.push(0);

    ms_1.get_min();
    ms_1.pop();
    ms_1.get_min();

    MinStack ms_2;
    std::cout << "ms_2" << std::endl;

    ms_2.push(-2);
    ms_2.push(0);
    ms_2.push(-3);

    ms_2.show();

    ms_2.get_min();
    ms_2.pop();
    ms_2.top();
    ms_2.get_min();

    MinStackB ms_3;
    std::cout << "ms_3" << std::endl;

    ms_3.push(0);
    ms_3.push(1);
    ms_3.push(0);

    ms_3.get_min();
    ms_3.pop();
    ms_3.get_min();

    MinStackB ms_4;
    std::cout << "ms_4" << std::endl;

    ms_4.push(-2);
    ms_4.push(0);
    ms_4.push(-3);

    ms_4.get_min();
    ms_4.pop();
    ms_4.top();
    ms_4.get_min();
}

File: neet_code_150/src/stack/min_stack.rs
#[derive(Debug)]
struct MinStack {
    volume: Vec<i32>,
    min_list: Vec<i32>,
}

// push, pop, top, get_minをO(1)で
// AC
// しかし、配列を二本使っているせいでメモリ効率が悪い
impl MinStack {
    fn new() -> Self {
        MinStack {
            volume: vec![],
            min_list: vec![],
        }
    }

    fn push(&mut self, val: i32) {
        if let Some(min_value) = self.min_list.iter().rev().next() {
            if val <= *min_value {
                self.min_list.push(val);
            }
        } else {
            self.min_list.push(val);
        }
        self.volume.push(val);
    }

    fn pop(&mut self) {
        match self.volume.pop() {
            Some(t) => {
                if let Some(min_value) = self.min_list.iter().rev().next() {
                    if t == *min_value {
                        match self.min_list.pop() {
                            Some(_t) => {}
                            None => {}
                        }
                    }
                }
            }
            None => {}
        }
    }

    fn top(&self) -> i32 {
        match self.volume.iter().rev().next() {
            Some(t) => {
                eprintln!("{t}");
                *t
            }
            None => panic!("empty"),
        }
    }

    fn get_min(&self) -> i32 {
        if let Some(min_value) = self.min_list.iter().rev().next() {
            eprintln!("{min_value}");
            *min_value
        } else {
            panic!("empty");
        }
    }
}

#[derive(Debug)]
struct MinStackC {
    volume: Vec<(i32, i32)>,
}

/*
struct Hoge {
    a: Vec<i32>,
    b: Vec<i32>,
}

よりも

struct Hoge {
    a: Vec<(i32, i32)>
}

の方がメモリ効率が良い？
*/

// push, pop, top, get_minをO(1)で
// AC
// こっちの方がメモリ効率が良かった
impl MinStackC {
    fn new() -> Self {
        MinStackC { volume: vec![] }
    }

    fn push(&mut self, val: i32) {
        if self.volume.is_empty() {
            self.volume.push((val, val));
            return;
        }

        if let Some(current) = self.volume.iter().rev().next() {
            let min_current = current.1;
            if val <= min_current {
                self.volume.push((val, val));
            } else {
                self.volume.push((val, min_current));
            }
        }
    }

    fn pop(&mut self) {
        match self.volume.pop() {
            Some(_current) => {}
            None => {}
        }
    }

    fn top(&self) -> i32 {
        match self.volume.iter().rev().next() {
            Some(top) => {
                eprintln!("{}", top.0);
                (*top).0
            }
            None => panic!("empty"),
        }
    }

    fn get_min(&self) -> i32 {
        match self.volume.iter().rev().next() {
            Some(top) => {
                //println!("{}", (*top).1);
                (*top).1
            }
            None => {
                panic!("empty");
            }
        }
    }
}

#[derive(Debug)]
struct MinStackAns {
    stack: Vec<i32>,
    min_stack: Vec<i32>,
}

impl MinStackAns {
    fn new() -> Self {
        Self {
            stack: vec![],
            min_stack: vec![],
        }
    }

    fn push(&mut self, val: i32) {
        self.stack.push(val);
        if self.min_stack.is_empty() || Some(&val) <= self.min_stack.last() {
            self.min_stack.push(val);
        }
    }

    fn pop(&mut self) {
        let val = self.stack.pop().unwrap();
        if Some(&val) == self.min_stack.last() {
            self.min_stack.pop();
        }
    }

    fn top(&self) -> i32 {
        self.stack.last().cloned().unwrap()
    }

    fn get_min(&self) -> i32 {
        self.min_stack.last().cloned().unwrap()
    }
}

// 時間を開けて解いたときの別解
// AC
struct MinStackLatter {
    data: Vec<(i32, i32)>,
}

impl MinStackLatter {
    fn new() -> Self {
        Self { data: vec![] }
    }

    pub fn push(&mut self, val: i32) {
        if let Some(stack_top) = self.data.last() {
            self.data.push((val, stack_top.1.min(val)));
        } else {
            self.data.push((val, val));
        }
    }

    pub fn pop(&mut self) {
        self.data.pop();
    }

    pub fn top(&mut self) -> i32 {
        self.data.last().unwrap().0
    }

    pub fn get_min(&self) -> i32 {
        self.data.last().unwrap().1
    }
}

fn main() {
    let mut ms_1 = MinStack::new();
    println!("ms_1");
    ms_1.push(-2);
    ms_1.push(0);
    ms_1.push(-3);

    ms_1.get_min();
    ms_1.pop();
    ms_1.top();
    ms_1.get_min();

    let mut ms_2 = MinStack::new();
    println!("ms_2");
    ms_2.push(0);
    ms_2.push(1);
    ms_2.push(0);

    ms_2.get_min();
    ms_2.pop();
    ms_2.get_min();

    let mut ms_4 = MinStackC::new();
    println!("ms_4");
    println!("ms_4: {:?}", ms_4);
    ms_4.push(0);
    println!("ms_4: {:?}", ms_4);
    ms_4.push(1);
    println!("ms_4: {:?}", ms_4);
    ms_4.push(0);
    println!("ms_4: {:?}", ms_4);

    ms_4.get_min();
    ms_4.pop();
    ms_4.get_min();

    let mut ms_5 = MinStackAns::new();
    println!("ms_5");
    println!("ms_5: {:?}", ms_5);
    ms_5.push(0);
    println!("ms_5: {:?}", ms_5);
    ms_5.push(1);
    println!("ms_5: {:?}", ms_5);
    ms_5.push(0);
    println!("ms_5: {:?}", ms_5);

    let mut s_0 = MinStackLatter::new();
    s_0.push(-2);
    s_0.push(0);
    s_0.push(-3);
    println!("s_0.get_min(): {}", s_0.get_min());
    s_0.pop();
    println!("s_0.top(): {}", s_0.top());
    println!("s_0.get_min(): {}", s_0.get_min());
}
